staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,272 @@
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.UntilWithoutRetry where premature termination is causing issues and flakes.
+// Please note that this is not resilient to etcd cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherClient       cache.Watcher
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+	doneChan            chan struct{}
+	hardFailDelay       time.Duration
+}
+
+// NewRetryWatcher creates a new RetryWatcher with default delay for hard failures.
+// See NewRetryWatcherWithDelay for details.
+func NewRetryWatcher(initialResourceVersion string, watcherClient cache.Watcher) (*RetryWatcher, error) {
+	return NewRetryWatcherWithDelay(initialResourceVersion, watcherClient, defaultHardFailDelay)
+}
+
+// NewRetryWatcherWithDelay creates a new RetryWatcher and allows specifying delay for hard failures.
+// It will make sure that watches gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watch method when first called.
+func NewRetryWatcherWithDelay(initialResourceVersion string, watcherClient cache.Watcher, hardFailDelay time.Duration) (*RetryWatcher, error) {
+	switch initialResourceVersion {
+	case "", "0":
+		// TODO: revisit this if we ever get WATCH v2 where it means start "now"
+		//       without doing the synthetic list of objects at the beginning
+		return nil, fmt.Errorf("initial RV %q is not supported due to issues with underlying WATCH, see #74022", initialResourceVersion)
+	default:
+		break
+	}
+
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherClient:       watcherClient,
+		stopChan:            make(chan struct{}),
+		doneChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+		hardFailDelay:       hardFailDelay,
+	}
+
+	go rw.receive()
+	return rw, nil
+}
+
+func (rw *RetryWatcher) send(event watch.Event) bool {
+	// Writing to an unbuffered channel is blocking operation
+	// and we need to check if stop wasn't requested while doing so.
+	select {
+	case rw.resultChan <- event:
+		return true
+	case <-rw.stopChan:
+		return false
+	}
+}
+
+// doReceive returns true when it is done, false otherwise.
+// If it is not done the second return value holds the time to wait before calling it again.
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
+	watcher, err := rw.watcherClient.Watch(metav1.ListOptions{
+		ResourceVersion: rw.lastResourceVersion,
+	})
+	// We are very unlikely to hit EOF here since we are just establishing the call,
+	// but it may happen that the apiserver is just shutting down (e.g. being restarted)
+	// This is consistent with how it is handled for informers
+	switch err {
+	case nil:
+		break
+
+	case io.EOF:
+		// watch closed normally
+		return false, 0
```

Comment:
anywhere this does `return false, 0`, we're still setting ourselves up for a hotloop, right?

Reply:
I think on the normal path, WATCH should be restarted immediately. 

What is a normal path could be the question. If you are unlucky to open a connection to apiserver that is just about to terminate, you likely want to start again by trying another watch as that wasn't error path you'd have to backoff on.

If your watch timeouts, you shouldn't sleep but try immediately.

I guess we could count consequent failures before getting to the processing phase but splitting failures into ones we want to retry immediately and ones we want to apply delay to seemed good enough for this use.

Reply:
> If your watch timeouts, you shouldn't sleep but try immediately.

why? reflectors delay for a second in exactly that scenario, so they don't hotloop:

https://github.com/kubernetes/kubernetes/blob/8d6f20ea4342327bb778a80c70c47ec9ad47e182/staging/src/k8s.io/client-go/tools/cache/reflector.go#L113

https://github.com/kubernetes/kubernetes/blob/8d6f20ea4342327bb778a80c70c47ec9ad47e182/staging/src/k8s.io/client-go/tools/cache/reflector.go#L133-L137

I think retryAfter should be a minimum of 1 second.

Reply:
To me, the difference is that informers are caches and this is a form of direct api call.

Caches are ok with a delay, I don't expect delays to be injected into a real-time api call on the happy path.

Reply:
but I think there is a way to address both concerns, let me update the PR...